pallets/proofs-dealer/src/lib.rs

```diff
@@ -603,7 +636,26 @@ pub mod pallet {
         ///
         /// Execution of this extrinsic should be refunded if the proof is valid.
         #[pallet::call_index(1)]
-        #[pallet::weight(Weight::from_parts(10_000, 0) + T::DbWeight::get().writes(1))]
+        #[pallet::weight({
+            let max_random_key_proofs = T::RandomChallengesPerBlock::get().saturating_mul(2u32.into());
+            let max_custom_key_proofs = T::MaxCustomChallengesPerBlock::get().saturating_mul(2u32.into());
+
+            let max_key_proofs = max_random_key_proofs.saturating_add(max_custom_key_proofs);
+
+            let key_proofs_len = SaturatedConversion::saturated_into::<u32>(
+                proof.key_proofs.len()
+            );
+            match key_proofs_len {
+                n if n <= max_random_key_proofs => {
+                    T::WeightInfo::submit_proof_no_checkpoint_challenges_key_proofs(n)
+                }
+                n if n <= max_key_proofs => {
+                    T::WeightInfo::submit_proof_with_checkpoint_challenges_key_proofs(n)
+                }
+                // More key proofs than `max_key_proofs` would inevitably fail the transaction.
+                n => T::WeightInfo::submit_proof_with_checkpoint_challenges_key_proofs(n),
+            }
+        })]
         pub fn submit_proof(
```

Comment:
note: we discussed having potentially 2 extrinsics that expect different input as proofs to have unique benchmarking for different challenge cases.

Reply:
Correct. We've setup a task to do that optimisation down the line.